driver/diagflat_driver.hpp

```diff
@@ -0,0 +1,390 @@
+
+    int RunForwardGPU() override;
+    int RunForwardCPU(); // Verify implements it
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU(); // Verify implements it
+
+    Tref GetTolerance();
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~DiagFlatDriver() override
+    {
+        miopenDestroyTensorDescriptor(outputTensor);
+        miopenDestroyTensorDescriptor(inputTensor);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
+    bool isContiguous;
+
+    miopenTensorDescriptor_t inputTensor;
+    miopenTensorDescriptor_t outputTensor;
+
+    std::unique_ptr<GPUMem> in_dev;
+    std::unique_ptr<GPUMem> out_dev;
+
+    std::vector<Tgpu> in;
+    std::vector<Tgpu> out;
+    std::vector<Tref> outhost;
+
+    int64_t offset;
+};
+
+template <typename Tgpu, typename Tref>
+int DiagFlatDriver<Tgpu, Tref>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int DiagFlatDriver<Tgpu, Tref>::GetandSetData()
+{
+    SetBNParametersFromCmdLineArgs();
+
+    std::vector<int> in_len = GetInputTensorLengthsFromCmdLine();
+    offset                  = inflags.GetValueInt("Offset");
+    isContiguous            = inflags.GetValueInt("contiguous") > 0 ? true : false;
+
+    auto inStride = ComputeStrides(in_len);
+    SetTensorNd(inputTensor, in_len, inStride, data_type);
+
+    auto input_numel = miopen::deref(inputTensor).GetElementSize();
+    auto outsz       = input_numel + abs(offset);
+    std::vector<int> out_len{outsz, outsz};
+
+    SetTensorNd(outputTensor, out_len, data_type);
+
+    return miopenStatusSuccess;
+}
+
+// Equivalent tensor.transpose(0, -1).contiguous().transpose(0, -1)
+template <typename Tgpu, typename Tref>
+std::vector<int> DiagFlatDriver<Tgpu, Tref>::ComputeStrides(std::vector<int> inputDim)
+{
+    if(!isContiguous)
+        std::swap(inputDim.front(), inputDim.back());
+    std::vector<int> strides(inputDim.size());
+    strides.back() = 1;
+    for(int i = inputDim.size() - 2; i >= 0; --i)
+        strides[i] = strides[i + 1] * inputDim[i + 1];
+    if(!isContiguous)
+        std::swap(strides.front(), strides.back());
+    return strides;
+}
+
+template <typename Tgpu, typename Tref>
+int DiagFlatDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only Forward (1) (Default = 1)", "int");
+    inflags.AddInputFlag("batchsize", 'n', "2", "Mini-batch size (Default=2)", "int");
+    inflags.AddInputFlag("in_channels", 'c', "5", "Number of Input Channels (Default=0)", "int");
+    inflags.AddInputFlag("in_d", 'D', "0", "Input Depth (Default=0)", "int");
+    inflags.AddInputFlag("in_h", 'H', "0", "Input Height (Default=0)", "int");
+    inflags.AddInputFlag("in_w", 'W', "4", "Input Width (Default=4)", "int");
```

Comment:
It's not necessary but using 4 flags to get the input tensor dim is quite verbose. Plz check other recent PRs (mine for ex) to see a better approach.